### DISCUSSION TITLE
Creating migration file without losing data in database

### DISCUSSION THREAD

**Discussion by ghost** [prisma/prisma#6367] [Q&A]
Hi everyone, 
so i was wondering almost every time i update my `schema.prisma` file and do `prisma migrate dev --name something` i get this warning. For now im using a second database as a place to do this migration , and then after the migration file is created i do `prisma migrate deploy` to apply it to my main database. My previous experience was using Sequelize, and there the migration doesn't require me to reset my database, which part am i missing or is there any different approach to solve this?

![prisma_lost_data](https://user-images.githubusercontent.com/33424694/113324525-674a1e00-9341-11eb-9484-30c4a053e31c.png)

**Comment by ghost** [comment]
@ryands17 hi Ryan i saw you answering one of my question before, mind taking a look at this one? really appreciate some help

**Reply by ryands17** [reply]
@peedo14 👋 

Could you share what was your previous schema and what are you trying to add in the new migration?

**Reply by ghost** [reply]
@ryands17  pretty much if i add new model/table or adding relation and foreign key, this will shows. it's more like, i wasn't really sure on how do migration properly in prisma. is there any workaround for this one? maybe seeding or something? i just don't want my data to lost and create/execute the migration smoothly

**Reply by ryands17** [reply]
I tried with a simple schema. First I added the `User` model then it's relation to the `Post` model and I am not facing any warning of data loss.

![image](https://user-images.githubusercontent.com/19697099/113713517-13ef1b80-9705-11eb-9c78-1abf26831b01.png)

Here's the sample project:

[test-prisma.zip](https://github.com/prisma/prisma/files/6264734/test-prisma.zip)

It would be great if you could provide a scenario where this happens.

**Reply by ghost** [reply]
i have an existing database, and i change a column name like this (rename column to haha in model)
```prisma
model AddressType {
  id          Int      @id @default(autoincrement())
  haha        String   @unique @db.VarChar(255)

  @@map("address_types")
}
```
then i run `npx prisma migrate dev --name testing`
<img width="1440" alt="Screen Shot 2021-04-06 at 20 08 01" src="https://user-images.githubusercontent.com/33424694/113715533-cd54ed80-9713-11eb-9aef-e8f12137a8e8.png">
then this happens, below that there's the notification asking to reset my database just like photo in the post

**Reply by ryands17** [reply]
In this case, after changing your column name, you can run the following command:

```
npx prisma migrate dev --create-only --name testing
```

You will still see the message but you can continue as nothing will be applied to the database. This will just create the `.sql` file corresponding to the change in `schema.prisma`. You need to then replace the contents of this `.sql` file with the `alter table` command to rename the column.

After you're done, run `npx prisma migrate dev`. This will just change the table name and there won't be any data loss.

**Reply by ghost** [reply]
@ryands17 will try to use it and update you with the result, anyway thanks for the time

**Reply by ghost** [reply]
@ryands17 actually same issues happens,i think my case might be a bit more complex to solve since there a lot of inconsistency in my db. one more thing though, can you explain to me what is lazy connect behaviour in prisma? and maybe about database pooling a bit as well as i feel like deploying to production im must aware of this in prisma. thanks

**Reply by ryands17** [reply]
@peedo14 It would be great if you could share the process that you did to change the column name in the table so that I can check.

> can you explain to me what is lazy connect behaviour in prisma

I'm not sure where you have heard this but I'm not aware of any such thing. Could you point me to where you have read this term?

> and maybe about database pooling a bit as well as i feel like deploying to production im must aware of this in prisma

When Prisma connects to the database, it creates a connection pool that makes a [specific amount of connections](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-management#connection-pool) to the database. If you are running on a long-running server, it will be fine, but for Serverless environments like Lambda, you need to set the `connection_limit` as low as possible [described here](https://www.prisma.io/docs/guides/deployment/deployment/#database-connection-limit).

**Reply by ghost** [reply]
@ryands17 all i did is just changing the name of the column as described before,for the migration this is more about my database in general i feel like. 

i'm currently trying do to transaction in my API, personally do you think transactions will slow down the whole application? im refering to [this](https://dzone.com/articles/practical-transaction-handling-in-microservice-arc) article.

about the prisma lazy connect behaviour, i've seen it here
<img width="716" alt="Screen Shot 2021-04-07 at 17 21 05" src="https://user-images.githubusercontent.com/33424694/113851410-aeaf2f00-97c5-11eb-9563-d4527377807c.png">

if i just instantiate prisma client on app start, do i still need database pooling? since only 1 connection is made to database right?

will read about the docs you linked as well

**Reply by ryands17** [reply]
Calling `connect` explicitly is not required and won't have any performance penalty as this happens only once in the entire application.

The `connection_limit` parameter needs to be set for the number of connections you want in the pool. This is mentioned in the docs I linked above. Otherwise, then connection is made based on the number of CPU's of the system you have deployed to. If you are running your application on Heroku or AWS EC2, then it's not required to add this and the default will be used.
